sklearn/tree/_utils.pyx

```diff
@@ -361,14 +362,8 @@ cdef class WeightedPQueue:
         # Resize if capacity not sufficient
         if array_ptr >= self.capacity:
             self.capacity *= 2
-            array = <WeightedPQueueRecord*> realloc(self.array_,
-                                                    self.capacity *
-                                                    sizeof(WeightedPQueueRecord))
-
-            if array == NULL:
-                # no free; __dealloc__ handles that
-                return -1
-            self.array_ = array
+            with gil:
+                safe_realloc(&self.array_, self.capacity)
```

Comment:
That's a stupid question (I'm not an expert in Cython :[ ): why do you need `with gil` here and not in [l346](https://github.com/scikit-learn/scikit-learn/pull/8002/files#diff-4df11ecb1ffdca75146bded11db520e4R346) ?